crates/compiler/src/lib.rs

```diff
@@ -1,7 +1,157 @@
+use cairo_m_compiler_diagnostics::{build_diagnostic_message, Diagnostic, DiagnosticSeverity};
+use cairo_m_compiler_parser::{parse_program, SourceProgram};
+use db::CompilerDatabase;
+
+/// Result type for compilation operations
+pub type Result<T> = std::result::Result<T, CompilerError>;
+
+/// Errors that can occur during compilation
+#[derive(Debug, Clone)]
+pub enum CompilerError {
+    /// Parse errors occurred
+    ParseErrors(Vec<Diagnostic>),
+    /// Semantic validation errors occurred
+    SemanticErrors(Vec<Diagnostic>),
+    /// MIR generation failed
+    MirGenerationFailed,
+    /// Code generation failed
+    CodeGenerationFailed(String),
+}
+
+impl std::fmt::Display for CompilerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::ParseErrors(diagnostics) => {
+                write!(f, "Parse errors: {} errors found", diagnostics.len())
+            }
+            Self::SemanticErrors(diagnostics) => {
+                write!(f, "Semantic errors: {} errors found", diagnostics.len())
+            }
+            Self::MirGenerationFailed => write!(f, "Failed to generate MIR"),
+            Self::CodeGenerationFailed(e) => write!(f, "Code generation failed: {}", e),
+        }
+    }
+}
+
+impl std::error::Error for CompilerError {}
+
+/// Options for compilation
+#[derive(Debug, Clone, Default)]
+pub struct CompilerOptions {
+    /// Enable verbose output
+    pub verbose: bool,
+}
+
+/// Compilation output including the compiled program and any diagnostics
+#[derive(Debug)]
+pub struct CompilerOutput {
+    /// The compiled program
+    pub program: Arc<CompiledProgram>,
+    /// Any warnings generated during compilation
+    pub warnings: Vec<Diagnostic>,
+}
+
+/// Compiles a Cairo-M source file from a string
+///
+/// # Arguments
+/// * `source_text` - The source code to compile
+/// * `source_name` - Name of the source file (for error reporting)
+/// * `options` - Compilation options
+///
+/// # Returns
+/// * `Ok(CompilerOutput)` - Successfully compiled program with any warnings
+/// * `Err(CompilerError)` - Compilation failed with errors
+pub fn compile_cairo(
+    source_text: String,
+    source_name: String,
+    options: CompilerOptions,
+) -> Result<CompilerOutput> {
+    let db = CompilerDatabase::new();
+    let source = SourceProgram::new(&db, source_text, source_name);
+
+    compile_from_source(&db, source, options)
+}
+
+/// Compiles a Cairo-M program from a SourceProgram
+///
+/// This is a lower-level API that allows reusing a database instance
+/// for incremental compilation scenarios.
+pub fn compile_from_source(
+    db: &CompilerDatabase,
+    source: SourceProgram,
+    _options: CompilerOptions,
+) -> Result<CompilerOutput> {
+    // Parse the program
+    let parsed_program = parse_program(db, source);
+
+    if !parsed_program.diagnostics.is_empty() {
+        return Err(CompilerError::ParseErrors(parsed_program.diagnostics));
+    }
```

Comment:
warnings are part of diagnostics no?
shouldn't there is a preprocessing phase to differentiate the 2?
Here if there is only warnings it will send back an error, it is the desired behavior?

Reply:
Oh ok, I was confused by the severity
What is the purpose of the severity of a diagnostics?
```rust
pub enum DiagnosticSeverity {
    Error,
    Warning,
    Info,
    Hint,
}

```

Reply:
The current implementation is correct. The parser only generates errors (not warnings), so returning an error for any diagnostics at line 95-97 is appropriate. The semantic phase (lines 102-117) properly separates errors from warnings:
- Parse errors: Always fail compilation
- Semantic errors: Fail compilation
- Semantic warnings: Collected and returned with the compiled output

No changes needed.

Reply:
Error => fails
Warning => Draw attention, but still compiles (unused var, etc), could be sub-optimal code
Info | Hints => informational, tbd (for now no such diag emitted)